File: cpp/Data_Structure/ListNode.cpp
#include <iostream>
using namespace std;

typedef int ElementType;

class ListNode {
public:
  struct Node {
    ElementType val;
    Node *next;
    Node(ElementType val) : val(val), next(nullptr) {}
    Node(Node *ptr) : val(ptr->val), next(ptr) {}
  };

  ListNode() {
    dummy = new Node(-1);
    size = 0;
    Min_Index = 0;
    Max_Index = 0;
  }

  ElementType get(int index) {
    if (index < 0 || index > (size - 1)) {
      return -1;
    }
    Node *cur = dummy->next;
    while (index--) {
      cur = cur->next;
    }
    return cur->val;
  }

  ElementType Get_Max() {
    Parse_Most_Value();
    Node *cur = dummy->next;
    while (Max_Index--) {
      cur = cur->next;
    }
    return cur->val;
  }

  ElementType Get_Min() {
    Parse_Most_Value();
    Node *cur = dummy->next;
    while (Min_Index--) {
      cur = cur->next;
    }
    return cur->val;
  }

  void ListNodeInsert(int index, ElementType val) {
    if (index < 0 || index > size) {
      return;
    }
    Node *cur = dummy;
    Node *newnode = new Node(val);
    while (index--) {
      cur = cur->next;
    }
    newnode->next = cur->next;
    cur->next = newnode;

    size++;
  }

  void ListNodeErase(int index) {
    if (index < 0 || index > (size - 1)) {
      return;
    }
    Node *cur = dummy;
    while (index--) {
      cur = cur->next;
    }
    Node *tmp = cur->next;
    cur->next = tmp->next;
    delete tmp;

    size--;
  }

  void HeadInsert(ElementType val) {
    Node *newnode = new Node(val);
    Node *head = dummy->next;
    dummy->next = newnode;
    newnode->next = head;

    size++;
  }

  void TailInsert(ElementType val) {
    Node *cur = dummy->next;
    while (cur) {
      cur = cur->next;
    }
    Node *newnode = new Node(val);
    newnode->next = cur->next;
    cur->next = newnode;

    size++;
  }

  void Print() {
    Node *cur = dummy->next;
    while (cur) {
      cout << cur->val << "->";
      cur = cur->next;
    }
    cout << "nullptr" << endl;
  }

private:
  int size;
  Node *dummy;
  int Max_Index;
  int Min_Index;

  void Parse_Most_Value() {
    Node *cur = dummy->next;
    for (int i = 0; i < size; i++) {
      Max_Index = max(Max_Index, cur->val);
      Min_Index = min(Min_Index, cur->val);
      cur = cur->next;
    }
  }
};

void test() {
  ListNode l;
  l.HeadInsert(1);
  l.ListNodeInsert(0, 0);
  l.ListNodeInsert(1, 1);
  l.ListNodeErase(2);
  l.Print();
  cout << l.get(1) << endl;
  cout << l.Get_Max() << endl;
  cout << l.Get_Min() << endl;
}

int main(int argc, char *argv[]) {
  test();
  return 0;
}

File: python/Data_Structure/ListNode.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None
        self.prev = None


class ListNode:
    def __init__(self, val):
        self.plist = Node(val)
        self.size = 1

    def insert(self, index, val):
        if index < 0 or index > self.size:
            return
        newnode = Node(val)
        cur = self.plist
        if index == 0:
            newnode.next = cur
            cur.prev = newnode
            newnode.prev = cur
            cur.next = newnode
            self.plist = newnode
        else:
            while index:
                cur = cur.next
                index -= 1
            pre = cur.prev
            pre.next = newnode
            newnode.prev = pre
            newnode.next = cur
            cur.prev = newnode
        self.size += 1

    def erase(self, index):
        if index < 0 or index > self.size:
            return
        elif index == 0:
            last = self.plist.prev
            next = self.plist.next
            next.prev = last
            last.next = next
            self.plist = next
        elif index == (self.size - 1):
            cur = self.plist.prev
            last = cur.prev
            last.next = self.plist
            self.plist.prev = last
        self.size -= 1

    def insert_head(self, val):
        self.insert(0, val)

    def insert_tail(self, val):
        self.insert(self.size, val)

    def erase_head(self):
        self.erase(0)

    def erase_tail(self):
        self.erase(self.size)

    def get(self, index):
        if index < 0 or index > self.size:
            return
        cur = self.plist
        while index:
            cur = cur.next
            index -= 1
        return cur.data

    def print(self):
        cur = self.plist
        for i in range(self.size):
            print(cur.data, end="->")
            cur = cur.next
        print("None")


if __name__ == "__main__":
    pl = ListNode(0)
    pl.insert_head(1)
    pl.insert_tail(2)
    pl.insert(3, 3)
    pl.print()
    pl.erase_tail()
    pl.print()
    pl.erase_head()
    pl.print()
    print(pl.get(3))
